go/plugins/dotprompt/dotprompt.go

```diff
@@ -321,3 +327,117 @@ func sortSchemaSlices(s *jsonschema.Schema) {
 		sortSchemaSlices(s.Items)
 	}
 }
+
+// WithTools adds tools to the prompt.
+func WithTools(tools ...ai.Tool) PromptOption {
+	return func(p *Prompt) error {
+		if p.Config.Tools != nil {
+			return errors.New("dotprompt.WithTools: cannot set tools more than once")
+		}
+
+		var toolSlice []ai.Tool
+		toolSlice = append(toolSlice, tools...)
+		p.Tools = toolSlice
+		return nil
+	}
+}
+
+// WithDefaultConfig adds default model configuration.
+func WithDefaultConfig(config *ai.GenerationCommonConfig) PromptOption {
+	return func(p *Prompt) error {
+		if p.Config.GenerationConfig != nil {
+			return errors.New("dotprompt.WithDefaultConfig: cannot set config more than once")
+		}
+		p.Config.GenerationConfig = config
+		return nil
+	}
+}
+
+// WithInputType uses the type provided to derive the input schema.
+// If passing eg. a struct with values, the struct definition will serve as the schema, the values will serve as defaults if no input is given at generation time.
+func WithInputType(input any) PromptOption {
+	return func(p *Prompt) error {
+		if p.Config.InputSchema != nil {
+			return errors.New("dotprompt.WithInputType: cannot set inputType more than once. Set either inputType or inputSchema")
+		}
+
+		p.Config.InputSchema = base.InferJSONSchemaNonReferencing(input)
+
+		// Set values as default input
+		inputDefault := base.SchemaAsMap(p.Config.InputSchema)
+		data, err := json.Marshal(input)
+		if err != nil {
+			return err
+		}
+
+		err = json.Unmarshal(data, &inputDefault)
+		if err != nil {
+			return err
+		}
+
+		p.Config.InputDefault = inputDefault
+		return nil
+	}
+}
+
+// WithOutputType uses the type provided to derive the output schema.
+func WithOutputType(output any) PromptOption {
+	return func(p *Prompt) error {
+		if p.Config.OutputSchema != nil {
+			return errors.New("dotprompt.WithOutputType: cannot set outputType more than once. Set either outputType or outputSchema")
+		}
+
+		p.Config.OutputSchema = base.InferJSONSchemaNonReferencing(output)
+		p.Config.OutputFormat = ai.OutputFormatJSON
+
+		return nil
+	}
+}
+
+// WithOutputFormat adds the desired output format to the prompt
+func WithOutputFormat(format ai.OutputFormat) PromptOption {
+	return func(p *Prompt) error {
+		if p.Config.OutputFormat != "" && p.Config.OutputFormat != format {
+			return errors.New("dotprompt.WithOutputFormat: outputFormat does not match set outputSchema")
```

Comment:
Nit: s/outputFormat/OutputFormat + the other ones to match the actual capitalization of the field in case someone goes looking for it.